### PR TITLE
cpu/stm32f1/dma: Initial DMA implementation for the STM32F1 family

### PR DESCRIPTION
**Pull request by DipSwitch**
As come to mind in #4952 we could implement DMA for the F1 family to make the MCU do other stuff or sleep while sending data over UART.

### CODE REVIEW

**Comment by DipSwitch** [comment]
Finished, with extra commenting for easy verification =)

File: cpu/stm32f1/include/periph_cpu.h [hunk 1]
```diff
@@ -154,6 +155,118 @@ void gpio_init_af(gpio_t pin, gpio_af_out_t af);
...
+     * DMA_ISR & DMA_IFCR have per channel 4 interrupt bits
+     * | b3      | b2     | b1       | b0     |
+     * |--------------------------------------|
+     * | TEIFx   | HTIFx  | TCIFx    | GIFx   |
+     * | Error   | Half   | Complete | Global |
+     *
+     * `2 << (channel * 4)` marks the TCIFx bit
+     */
+
+    if (channel < 7)
```
Comment (DipSwitch) [inline cpu/stm32f1/include/periph_cpu.h]: if (channel >= 7)
channel -= 7

File: cpu/stm32f1/include/periph_cpu.h [hunk 2]
```diff
@@ -154,6 +155,118 @@ void gpio_init_af(gpio_t pin, gpio_af_out_t af);
...
+     * | 7    | DMA2_Channel1_IRQn  | 56    |
+     * | 8    | DMA2_Channel2_IRQn  | 57    |
+     * | 9    | DMA2_Channel3_IRQn  | 58    |
+     * | 10-11| DMA2_Channel4_5_IRQn| 59    |
+     */
+
+    if (channel < 7) {
+        NVIC_EnableIRQ((IRQn_Type)((int)DMA1_Channel1_IRQn + channel));
+    }
+    else if (channel < 10) {
```
Comment (DipSwitch) [inline cpu/stm32f1/include/periph_cpu.h]: < 11

**Comment by DipSwitch** [comment]
About the abstraction @lebrush was talking about, STM has 2 types of DMA peripherals. The one (channel based) used in the F1 and the one used in the others (stream / channel with matrix interconnect)

**Comment by kYc0o** [comment]
Can you please rebase? This will be reviewed for the next release.

**Comment by DipSwitch** [comment]
This weekend I'll rebase all and see what can get fixed. Buzy times
unfortunaly!

**Comment by miri64** [comment]
Postponed due to feature freeze (still needs rebase though ;-))

**Comment by OlegHahm** [comment]
@DipSwitch, any chance that you will pick this up again any time soon?

**Comment by lebrush** [comment]
I'm working in a solution for all stm

**Comment by miri64** [comment]
No significant progress for over a year and @vincent-d seems to have picked up the work. Closing as memo for now.

**Pull request closed without merging**